File: libs/sdk/typescript/src/sdk.ts
import * as yaml from "yaml";
import * as schema from "./schema";

type PipelineStep =
    | schema.CommandStep
    | schema.WaitStep
    | schema.InputStep
    | schema.TriggerStep
    | schema.GroupStep;

export class Pipeline {
    private steps: PipelineStep[] = [];

    constructor() {
        this.steps = [];
    }

    addStep(step: PipelineStep) {
        this.steps.push(step);
        return this;
    }

    toJSON() {
        return JSON.stringify(
            {
                steps: this.steps,
            },
            null,
            4
        );
    }

    toYAML() {
        return yaml.stringify({
            steps: this.steps,
        });
    }
}

File: libs/sdk/go/sdk.go
package buildkite

import (
	"encoding/json"
)

type Pipeline struct {
	steps []Step
}

func (p *Pipeline) AddStep(step Step) {
	p.steps = append(p.steps, step)
}

func (p *Pipeline) ToJSON() (string, error) {
	data, err := json.Marshal(p.steps)
	if err != nil {
		return "", err
	}
	return string(data), nil
}
